File: src/engine/graphics/opengl/opengl_instance.cpp
#include "opengl_instance.h"
#include <gk_types_lib/job/job_thread.h>
#include <glad/glad.h>

graphics::OpenGLInstance* graphics::OpenGLInstance::init(gk::JobThread* renderThread)
{
	auto future = renderThread->runJob(gladLoadGL);
	int result = future.wait();
	if (result == GL_FALSE) {
		std::cout << "Failed to load OpenGL" << std::endl;
		exit(EXIT_FAILURE);
	}
	
	return new OpenGLInstance();
}

void graphics::OpenGLInstance::clear()
{
	glClear(GL_COLOR_BUFFER_BIT);
}

File: src/engine/graphics/opengl/OpenGLInstance.zig
const std = @import("std");
const c = @import("../../clibs.zig");
const assert = std.debug.assert;
const JobThread = @import("../../types/job_system.zig").JobThread;
const Engine = @import("../../Engine.zig");

const Self = @This();

pub fn init(renderThread: *JobThread) Self {
    const future = renderThread.runJob(c.gladLoadGL, .{}) catch unreachable;
    const result: c_int = future.wait();
    if (result == c.GL_FALSE) {
        @panic("Failed to load OpenGL");
    }

    return Self{};
}

/// # Asserts
///
/// Can only be called the current engine instance's render thread.
pub fn clear() void {
    assert(Engine.isCurrentOnRenderThread());
    c.glClear(c.GL_COLOR_BUFFER_BIT);
}
